### PR TITLE
(maint) Generate .gemspec

### PR DESCRIPTION
**Pull request by kylog**

### CODE REVIEW

**Comment by kylog** [comment]
Note that this drops the generated `.gemspec` in the root for parity with where .gemspec is found in actual ruby gem projects like puppet and hiera. This allows https://github.com/puppetlabs/puppet-agent/pull/345 to use consistent boilerplate for installing the gemspec for each of puppet/hiera/facter.

**Comment by MikaelSmith** [comment]
What does it do with the gemspec?

**Comment by kylog** [comment]
puppet-agent installs it to meet the needs of https://tickets.puppetlabs.com/browse/CPR-190.

File: lib/.gemspec.in [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  mdata = version.match(/(\d+\.\d+\.\d+)/)
+  s.version = mdata ? mdata[1] : version
+
+  s.required_rubygems_version = Gem::Requirement.new(">= 0")
+  s.authors = ["Puppet Labs"]
+  s.date = "2012-08-08"
+  s.description = "You can prove anything with facts!"
+  s.email = "info@puppetlabs.com"
+  s.executables = ["facter"]
+  s.files = ["bin/facter"]
```
Comment (joshcooper) [inline lib/.gemspec.in]: Does this path exist once the faux facter gem is "installed"?

Comment (kylog) [inline lib/.gemspec.in]: That's a really good question and to be honest I'm not sure what the contract really is for this field (or most of the others tbh). The docs for this field are at http://guides.rubygems.org/specification-reference/#files but again the exact semantics frankly aren't that clear to me.

Thoughts on what fields we actually *need* to provide here? I'd really rather pare this down to what is actually needed.

Btw, if we agree on a set, we may want to apply these lessons to the actual gem projects (puppet and hiera) as well. I already applied some fixup to the puppet gemspec in https://github.com/puppetlabs/puppet/pull/4223, but I suspect there's more cargo cult to be purged, both in a real gem install and in a puppet-agent install.

Comment (MikaelSmith) [inline lib/.gemspec.in]: I don't have a problem with this the way it is, it's as true os Hiera or Puppet's Gemspec.

Comment (MikaelSmith) [inline lib/.gemspec.in]: Looking at the docs, `executables` isn't needed and is probably wrong (as it says all files are expected to be in bindir and must be executable Ruby files). `files` is required; it's supposed to be all files you can `require`, which kind of makes sense as `facter`/`hiera`/`puppet`, so this seems ok.

**Comment by kylog** [comment]
I removed the `executables` line from `.gemspec.in`.

File: lib/.gemspec.in [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+Gem::Specification.new do |s|
+  s.name = "facter"
+  version = "@LIBFACTER_VERSION_MAJOR@.@LIBFACTER_VERSION_MINOR@.@LIBFACTER_VERSION_PATCH@"
+  mdata = version.match(/(\d+\.\d+\.\d+)/)
+  s.version = mdata ? mdata[1] : version
+
+  s.required_rubygems_version = Gem::Requirement.new(">= 0")
+  s.authors = ["Puppet Labs"]
+  s.date = "2012-08-08"
```
Comment (stahnma) [inline lib/.gemspec.in]: Should we have a more updated date? (nonblocking)

Can we specify a min ruby version above 2?

**Comment by MikaelSmith** [comment]
I'm fine with this.

**Pull request merged**